holonix/README.md

```diff
@@ -0,0 +1,128 @@
+- Dependencies are hashed
+  - "Dependency hell" is avoided
+  - Nice parallels with the hashed zomes model
+  - Security + reliability benefits
+- Dependencies can be garbage collected with the `nix-collect-garbage` command
+- Single "package manager" across most operating systems
+- Ability to ship utility scripts in the `shell.nix` file
+- Access to the nix functional programming language for dependencies/script management
+  - Allows for a structured approach to coding the ops infrastructure
+  - Modern programming paradigm such as immutability, scopes, data structures, etc.
+- NixOS runs on HoloPorts so `nix-shell` provides similar behaviour/environment
+  - Everything is pinned to the same nix channel and commit
+
+And other benefits:
+
+- Active and helpful community of contributors
+  - IRC at #nixos
+  - wiki at https://nixos.wiki/
+- Ability to bundle scripts into convenient binaries for daily tasks
+- Nice CLI interface with many flexible, well documented configuration options
+- Available in shebang form as `#! /usr/bin/env nix-shell`
+  - http://iam.travishartwell.net/2015/06/17/nix-shell-shebang/
+
+## Installation
+
+Follow the instructions to install `nix-shell` from:
+
+https://nixos.org/nix/download.html
+
+Once installed run `nix-shell` from the `holochain-rust` repository root.
+The `nix-shell` command will detect the `shell.nix` file and build everything.
+
+The first run will be slow as nix downloads, builds and caches dependencies.
+Everything is downloaded to `/nix/store/...` and re-used on subsequent runs.
+
+## Contributing
+
+The structure of our nix derivations is designed to be as modular as possible.
+
+The folder structure is a basic heirarchy, something like:
+
+```
+holonix
+ |_github
+ |  |_config.nix
+ |_nixpkgs
+ | |_nixpkgs.nix
+ |_node
+ | |_src
+ | | |_flux.nix
+ | | |_...
+ | |_build.nix
+ |_...
+shell.nix
+```
+
+The `shell.nix` file is used by `nix-shell` automatically by default.
+This consumes `holonix/**` and does not provide any new derivations.
+End-users should not need to interact with `holonix/**` outside `shell.nix`.
+
+There are a few basic convensions to follow:
+
+- Nest folders according to theme/tech/specificity
+  - e.g. conductor management for conductor `x` sits under `conductor/x/**`
+- All configuration strings and other primitives sit in a local `config.nix`
+- Structure configuration as nested `foo.bar` rather than `foo-bar`
+  - e.g. `holonix/release/config.nix` has a few good examples of this
+- All used and generated build inputs sit in a local `build.nix`
+  - `build.nix` files should "bubble up" to the root one level at a time
+    - e.g. `build.nix` imports `conductor/build.nix` imports `conductor/node/build.nix`
+  - root `build.nix` and `src` should only aggregate deeper derivations
+- Scripts for binaries sit in named `foo.nix` files under `thing/src/foo.nix`
+  - There is standard boilerplate for this, see an existing file for examples
```

Comment:
Can you point to an actual example file here?  "thing/src/foo.nix" is so general.  What's the "thing"?  the binary?  Or is "foo" the binary?

Reply:
oooh, see i'm talking about `writeShellScriptBin` (see below) where nix treats the shell script as a binary

this doesn't refer to binaries like `holochain` and `hc`, it just means that if you had a "binary" command called `hc-thing-foo` you'd put it on the file path `thing/src/foo.nix`